File: Chapter9_Example/src/main/java/EventDispatcher.java
import java.util.List;

public class EventDispatcher {
    private final MessageBus _messageBus;
    private final IDomainLogger _domainLogger;

    public EventDispatcher(
            MessageBus messageBus,
            IDomainLogger domainLogger) {
        _messageBus = messageBus;
        _domainLogger = domainLogger;
    }

    public void dispatch(List<IDomainEvent> events) {
        for(IDomainEvent ev : events) {
            dispatch(ev);
        }
    }

    private void dispatch(IDomainEvent ev) {
        switch (ev) {
            case EmailChangedEvent emailChangedEvent ->
                    _messageBus.sendEmailChangedMessage(emailChangedEvent.userId(), emailChangedEvent.newEmail());
            case UserTypeChangedEvent userTypeChangedEvent ->
                    _domainLogger.userTypeHasChanged(userTypeChangedEvent.userId(), userTypeChangedEvent.oldType(), userTypeChangedEvent.newType());
            default -> throw new IllegalArgumentException("지원하는 이벤트 타입이 아닙니다.");
        }
    }
}

File: unit-testing/src/main/kotlin/book/chapter10/unitofwork/EventDispatcher.kt
package book.chapter10.unitofwork

class EventDispatcher(
    private val messageBus: IMessageBus,
    private val domainLogger: IDomainLogger,
) {

    fun dispatch(events: List<IDomainEvent>) {
        for (event in events) {
            dispatch(event)
        }
    }

   private fun dispatch(event: IDomainEvent) {
        when(event) {
            is EmailChangedEvent -> messageBus.sendEmailChangedMessage(event.userId, event.newEmail)
            is UserTypeChangedEvent -> domainLogger.userTypeHasChanged(event.userId, event.oldType, event.newType)
        }
    }
}